Lazy Template Loading - Nukie/Wiz

ZephyrTFA opened the pull request:
## About The Pull Request

Removes the nukie base and wizard den from the base centcom map. They are instead now lazy loaded as required.
To make a new lazy load template is fairly simple, make a map, allocate an area for it, and place a marker at the bottom left corner of that area. I have it check an area to ensure that if someone makes the map larger than expected but doesn't account for the template allocation it doesn't overwrite stuff without warning

[Replaces some improper CHECK_TICKs with MAPLOADING_CHECK_TICKs](https://github.com/tgstation/tgstation/pull/71785/commits/c7fbca9148812b392dca76d5450fb6bb86716462)

Atom init has already been tripped by the time we get to this portion of the loading, so if we don't use the right check tick, we will potentially block unrelated init attempts. This is bad. (Lemon edit, I want this in the commit desc)

## Why It's Good For The Game

Cuts down on init times.
Closes https://github.com/tgstation/dev-cycles-initiative/issues/17
## Changelog
:cl:
admin: New mapping verb to load lazy templates as needed. In your admin tab under the Mapping category.
/:cl:
this isnt technically player visible, so not sure it needs a changelog

ZephyrTFA committed:
calculate template sizes correctly

ZephyrTFA committed:
lazy template loading

ZephyrTFA committed:
dont forget recovery

san7890 commented on code/controllers/subsystem/mapping.dm:
```diff
@@ -335,6 +338,7 @@ Used by the AI doomsday and the self-destruct nuke.
 
 	z_list = SSmapping.z_list
 	multiz_levels = SSmapping.multiz_levels
+	loaded_lazy_templates = SSmapping.lazy_load_template
```
```suggestion
	loaded_lazy_templates = SSmapping.lazy_load_templates
```

yo code ain't compiling

Mothblocks commented:
Need to measure how much this actually saves, and how noticable the init for players is when nukies roll

Mothblocks commented:
Haven't checked yet but make sure that it loads when hijacked, but also that an admin can explicitly load it. Admins including myself have used the nuke ops base to prepare for a hijack we know is coming.

ZephyrTFA committed:
Update code/controllers/subsystem/mapping.dm

Co-authored-by: san7890 <the@san7890.com>

ZephyrTFA committed:
Update mapping.dm

ZephyrTFA committed:
Update emergency.dm

ZephyrTFA commented:
Few more things to touch up

JohnFulpWillard commented:
can we do this for abductors too (it doesn't really change much for init times but I think it would be nice to be consistent)

Mothblocks commented:
I'll profile this when you have it passing CI

LemonInTheDark commented on code/_globalvars/lists/mapping.dm:
```diff
@@ -118,6 +118,7 @@ GLOBAL_LIST_EMPTY(deathsquadspawn)
 GLOBAL_LIST_EMPTY(emergencyresponseteamspawn)
 GLOBAL_LIST_EMPTY(ruin_landmarks)
 GLOBAL_LIST_EMPTY(bar_areas)
+GLOBAL_LIST_EMPTY(lazy_template_pivots)
```
this could be key -> pivot

LemonInTheDark replied:
should doc the format

LemonInTheDark commented on code/controllers/subsystem/mapping.dm:
```diff
@@ -844,6 +848,22 @@ GLOBAL_LIST_EMPTY(the_station_areas)
+		return
+	LAZYADDASSOC(loaded_lazy_templates, template_key, TRUE)
+
+	var/obj/effect/landmark/lazy_template_pivot/found
+	for(var/obj/effect/landmark/lazy_template_pivot/lazy_template_pivot in GLOB.lazy_template_pivots)
+		if(lazy_template_pivot.key != template_key)
+			continue
+		found = lazy_template_pivot
+		break
+
```
would yeet all of this

LemonInTheDark replied:
that or this should be a "load each of these pivots" pattern, rather then loading just one

LemonInTheDark commented on code/game/objects/effects/landmarks.dm:
```diff
@@ -443,6 +443,67 @@ INITIALIZE_IMMEDIATE(/obj/effect/landmark/start/new_player)
+			CRASH("Attempted to load a lazy template pivot larger than the allocation for the pivot")
+
+	// change the area of the turfs to space
+	var/area/space_area = GLOB.areas_by_type[/area/space]
+	for(var/turf/turf as anything in affected)
+		var/area/turf_area = get_area(turf)
+		turf_area.turfs_to_uncontain += turf
+		space_area.contents += turf
+		space_area.contained_turfs += turf
+		turf.baseturfs = /turf/baseturf_bottom
```
why move the turfs to space?

ZephyrTFA replied:
Didn't want the area to persist

ZephyrTFA replied:
Which I now realize is stupid

LemonInTheDark commented on code/game/objects/effects/landmarks.dm:
```diff
@@ -443,6 +443,67 @@ INITIALIZE_IMMEDIATE(/obj/effect/landmark/start/new_player)
+	var/list/affected = loading.get_affected_turfs(my_turf)
+	for(var/turf/turf as anything in affected)
+		var/area/turf_area = get_area(turf)
+		if(turf_area.type != /area/misc/lazy_pivot_allocation)
+			CRASH("Attempted to load a lazy template pivot larger than the allocation for the pivot")
```
could we acomplish this with a variable on the landmark? loading non default space turfs is slower then just default ones

ZephyrTFA replied:
not without losing a sanity guard check against mappers increasing the size of a template without updating the pivot

LemonInTheDark replied:
compare the size of the map to the size defined on the thing. a feature of the maploader is being able to dryload map files to get an idea of their size before placing them in the world

LemonInTheDark commented on code/controllers/subsystem/mapping.dm:
```diff
@@ -844,6 +848,22 @@ GLOBAL_LIST_EMPTY(the_station_areas)
 
 			true_to_offset_planes[string_real] |= offset_plane
 
+/datum/controller/subsystem/mapping/proc/lazy_load_template(template_key)
```
how appropriate is sleeping for like 10 seconds in the different places this is used?
loading post init will always take longer then loading pre init, because of changeturf issues

ZephyrTFA replied:
I've noticed it's not actually taking very long to load the map. It's almost instant, locally atleast

LemonInTheDark replied:
interesting

ZephyrTFA committed:
Update mapload_space_verification.dm

ZephyrTFA committed:
Update landmarks.dm

ZephyrTFA committed:
Update mapping.dm

ZephyrTFA committed:
Update outfits.dm

ZephyrTFA commented:
now also loads the nukie base if you select equip someone into nukie

ZephyrTFA committed:
Update mapping.dm

ZephyrTFA committed:
g

ZephyrTFA committed:
admin template loading
use cached boundary data instead of iter' over turfs

ZephyrTFA committed:
wrong copy/paste

ZephyrTFA commented:
alright reviews addressed and stuff works

LemonInTheDark commented on code/modules/admin/verbs/mapping.dm:
```diff
@@ -483,3 +484,25 @@ GLOBAL_VAR_INIT(say_disabled, FALSE)
+	if(!check_rights(R_SPAWN))
+		return
+
+	var/choice = tgui_input_list(usr, "Key?", "Lazy Loader", LAZY_TEMPLATE_KEY_LIST_ALL)
```
would be nice to have human readable names for these. actually  that could just BE the key but anyhow

ZephyrTFA replied:
I'm personally using the format LT_NAME_OF_TEMPLATE

LemonInTheDark replied:
yes but it'd be nice to give admins like, proper names

ZephyrTFA commented on code/modules/admin/verbs/mapping.dm:
```diff
@@ -483,3 +484,25 @@ GLOBAL_VAR_INIT(say_disabled, FALSE)
+	if(!choice)
+		return
+
+	if(!length(GLOB.lazy_template_pivots))
```
```suggestion
	if(!length(GLOB.lazy_template_pivots[choice]))
```

ZephyrTFA committed:
Update code/modules/admin/verbs/mapping.dm

ZephyrTFA committed:
Update mapping.dm

Mothblocks commented:
This saves about 0.4-0.7s. I was hoping for a lot more...but total init time for me is now about 33.4s when I merge master on top of this, so I'm happy as long as this is maintainable and a good framework. Still haven't looked at code tho.

LemonInTheDark commented:
Can we do this to the ninja room too?

ZephyrTFA committed:
Update lazy_templates.dm

ZephyrTFA committed:
Update mapping.dm

Mothblocks commented:
also can you please explain what makes this system different from map_template?

ZephyrTFA commented:
> also can you please explain what makes this system different from map_template?

what do you mean by this, it uses map templates to load stuff?

ZephyrTFA commented:
> i would like to know the answers to the "can we" questions for the ninja room and abductors pod, but i dont think its necessary to put in this pr if its not trivial

its easy to do, just minorly time consuming; give me like an hour

ZephyrTFA committed:
lazy load ninja holding

ZephyrTFA committed:
multi allocation templates
abductor ships

Mothblocks commented:
I guess what I'm asking is why is this a framework intruding in on SSmapping that also adds its own landmark and not *just* a map template? The SSmapping thing could just be to have a centralized place to pull these, and that is fine, but the landmark stuff weirds me out

ZephyrTFA commented:
> I guess what I'm asking is why is this a framework intruding in on SSmapping that also adds its own landmark and not _just_ a map template? The SSmapping thing could just be to have a centralized place to pull these, and that is fine, but the landmark stuff weirds me out

There needs to be a landmark to actually signal where the lazy templates are going to be loaded at on the map itself. Otherwise there would be no way to locate where the template should be loaded

ZephyrTFA commented:
> Why do we need to load them anywhere in particular, though? We don't care that the nuke base or wizard den is on CentCom, they're only there for historical reasons. I guess stuff checks is_centcom_level now but it just feels ready off, like we're going to eventually change the nuke ops base and then suddenly it eats into the range of other stuff.

Do you want me to add an empty Z level dedicated to lazy templates? And it'll keep track of used space accordingly?

ZephyrTFA commented:
Probably need a subsystem to manage it though, SSlazy_mapping

ZephyrTFA committed:
Merge branch 'master' into lazyantagbase

ZephyrTFA committed:
Update lazy_templates.dm

ZephyrTFA committed:
Update mapping.dm

ZephyrTFA commented:
> > Probably need a subsystem to manage it though, SSlazy_mapping
> 
> You should just use the reservations system, the one that divides up empty z levels into usable space for loading maps

mothblocks wants me to have them be lazy maps instead, each lazy map will get a dedicated z level

LemonInTheDark commented:
could honestly still use reservations for that (Tho a whole z level seems a bit overkill, and would for sure increase load times)

Mothblocks commented:
I don't think they need to be a separate z-level, I just think they don't need to be in specific points at CentCom

ZephyrTFA commented:
> I don't think they need to be a separate z-level, I just think they don't need to be in specific points at CentCom

How exactly do you expect me to accomplish finding a place to load them dynamically without being incredibly inefficient and bad

ZephyrTFA committed:
Merge remote-tracking branch 'upstream/master' into lazyantagbase

ZephyrTFA committed:
refactor into datums

ZephyrTFA committed:
move it to being an admin verb

ZephyrTFA committed:
add some create/destroy logic

ZephyrTFA committed:
if the template is deleted, remove from the global list, ideally this never happens

Fikou commented:
so do they no longer spawn on centcom? thats disappointing.

Mothblocks commented:
Yeah but also if they did what I asked and I think they did you'll be able to force them to spawn as an admin

Going to TM this tonight

Mothblocks commented:
It really doesn't work at all. I spawned nuke ops and it spawned the base, but of the four nukies, only two actually showed up, were in their normal job outfits (and according to one, they were on station then teleported, even though there should be no delay), and the other two were in this weird limbo where they were stuck on the Join Game screen but were told they were nukies and I had to make new bodies and give them keys

ZephyrTFA commented:
> It really doesn't work at all. I set nuke ops for roundstart on 196048 and it spawned the base, but of the four nukies, only two actually showed up, were in their normal job outfits (and according to one, they were on station then teleported, even though there should be no delay), and the other two were in this weird limbo where they were stuck on the Join Game screen but were told they were nukies and I had to make new bodies and give them keys

I'll look into it, probably missed something somewhere.

ZephyrTFA committed:
forgot to make sure nuke base is loaded when we start round

Mothblocks commented:
will check tmrw when i can poop another nuke round

ZephyrTFA commented:
abductors don't spawn correctly, ruleset needs to lazy load before pulling spawn points

Mothblocks commented:
round 196251, zephyr joined a test merge

abductor spawned i  believe it was the agent on arrivals but the scientist back at home base, roundstart nuke ops worked fine (next time we'll test with midrounds), wizard didn't spawn because it couldn't find a point, but force spawning worked fine

ZephyrTFA committed:
ensure templates get loaded for rulesets

Mothblocks commented:
Also put in your changelog the new admin verb for spawning the stuff, it can be useful for pre-emptive hijacks

san7890 commented:
Forgot about this PR and merged #72014. My bad.

Mothblocks commented:
its fine the problem is tooling not merging

san7890 committed:
Merge remote-tracking branch 'upstream/master' into pr/71785

san7890 committed:
fixes up the merge conflict

zephyr asked if i could do this so i did it

takes steps to prevent reverting 9cb866896a7b646f1b596a4d3a0136ecbdccae7a

Mothblocks commented:
its not spawning roundstart wizards

ZephyrTFA committed:
Update dynamic_rulesets_roundstart.dm

Mothblocks commented:
Eeeeee, causing some weird ass issues where we think it's conflicting with the arrivals shuttle on 196431

![image](https://user-images.githubusercontent.com/35135081/208212396-0d7fe0b9-ffdb-4175-ad60-83a287f72eda.png)

![image](https://user-images.githubusercontent.com/35135081/208212403-bfffc146-bae2-4ca1-b276-469589538c20.png)

![image](https://user-images.githubusercontent.com/35135081/208212405-77b458cf-a4e0-47bf-8c6a-6f39c9b384a9.png)

Mothblocks commented on code/__HELPERS/_lists.dm:
```diff
@@ -1099,3 +1099,6 @@
 
 		last_value = value
+
+/proc/coords2turf(list/coords)
```
Document

Mothblocks commented on code/__HELPERS/lazy_templates.dm:
```diff
@@ -0,0 +1,7 @@
+/proc/gen_lazy_template_map()
```
Document

`gen` -> `generate`

Mothblocks commented on code/__HELPERS/lazy_templates.dm:
```diff
@@ -0,0 +1,7 @@
+/proc/gen_lazy_template_map()
+	. = list()
```
`return .` at the end

Mothblocks commented on code/datums/lazy_template.dm:
```diff
@@ -0,0 +1,76 @@
+
+/datum/lazy_template/Destroy(force, ...)
+	if(!force)
+		return
```
```suggestion
		stack_trace("Something is trying to delete [type]")
		return QDEL_HINT_LETMELIVE
```

Mothblocks commented on code/datums/lazy_template.dm:
```diff
@@ -0,0 +1,76 @@
+
+	var/load_path = "[map_dir]/[map_name].dmm"
+	if(uses_multiple_allocations)
+		var/times = LAZYACCESS(multiple_allocation_hash, type) || 0
```
I wouldn't bother with the lazy list here and would just default to `list()`, makes the rest of this uglier

Mothblocks commented on code/game/area/areas/misc.dm:
```diff
@@ -34,3 +34,10 @@
 	name = "Test Room"
 	icon_state = "test_room"
+
+/area/misc/lazy_pivot_allocation
```
Do we still need this? I assume we are just loading map templates

Mothblocks commented on code/modules/mapping/map_template.dm:
```diff
@@ -152,9 +152,9 @@
 	if(!T)
 		return
-	if(T.x+width > world.maxx)
+	if((T.x+width) - 1 > world.maxx)
```
What's this?

ZephyrTFA replied:
fixing the calculations for whether or not attempting to load a map at a specificed coord will go out of bounds.
width is one based not zero based so you need to subtract one from the calculation

ZephyrTFA committed:
not needed anymore

ZephyrTFA committed:
doc and assist for coords2turf

ZephyrTFA committed:
doc and rename generate_lazy_templates

ZephyrTFA committed:
forgot to change this call

ZephyrTFA committed:
unused define

ZephyrTFA committed:
return at end

ZephyrTFA committed:
some code doc for lazy_load

ZephyrTFA committed:
Update code/datums/lazy_template.dm

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

ZephyrTFA committed:
remove the old area from mapload verif

ZephyrTFA committed:
Merge branch 'lazyantagbase' of github.com:ZephyrTFA/tgstation into lazyantagbase

Mothblocks commented:
Happened again with a TM to get #72064, we gotta figure this out

LemonInTheDark commented:
For context:
the shuttle's area is not being initialized. similarly, the shuttle's turfs aren't either. atoms are fine.

I have no idea why this is happening. Second time it happen it was a nukie round. Not sureif the first had any lateloading

Mothblocks commented:
196507 didn't seem to have any nukies though :(

ZephyrTFA committed:
Merge remote-tracking branch 'upstream/master' into lazyantagbase

JohnFulpWillard commented:
Nukie shuttle not loading breaks Syndicate headsets for traitors because they don't have a radio server

Mothblocks commented on code/__DEFINES/lazy_templates.dm:
```diff
@@ -0,0 +1,13 @@
+#define LAZY_TEMPLATE_KEY_NINJA_HOLDING_FACILITY "LT_NINJAHOLDING"
+#define LAZY_TEMPLATE_KEY_ABDUCTOR_SHIPS "LT_ABDUCTORSHIPS"
+
+#define LAZY_TEMPLATE_KEY_LIST_ALL list( \
```
Nit:

```suggestion
#define LAZY_TEMPLATE_KEY_LIST_ALL(...) list( \
```

...or make it a global list. Creating a list is not free or constant so it shouldn't be implied that it is.

Mothblocks commented on code/__DEFINES/lazy_templates.dm:
```diff
@@ -0,0 +1,13 @@
+	"Abductor Ships" = LAZY_TEMPLATE_KEY_ABDUCTOR_SHIPS, \
+)
+
+GLOBAL_LIST_INIT(lazy_templates, generate_lazy_template_map())
```
`__DEFINES` should only be defines, I know a lot of stuff does this right now but it annoys the hell out of me and regularly gets in the way. This should go either inside a `_globals` file, or with the rest of lazy template code.

LemonInTheDark committed:
Adds logging for an atom not initing post master init, along with atom init being toggled

LemonInTheDark committed:
Replaces some improper CHECK_TICKs with MAPLOADING_CHECK_TICKs

Atom init has already been tripped by the time we get to this portion of
the loading, so if we don't use the right check tick, we will
potentially block unrelated init attempts. This is bad.

Addust commented:
what about NTnet's Wirecarp app? it has already proven able to remotely detect what White Ship it is, and if it sees the nukie base network i'm pretty sure any competent RD/captain/greyshit with a stolen ID would report it over radio

Mothblocks commented:
can you elaborate on that? i dont know anything about wirecarp

also @LemonInTheDark has a fix for the blocking bug on this that they need to pr to the branch

ZephyrTFA committed:
remove debug code

ZephyrTFA committed:
turn define into macro

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

ZephyrTFA committed:
moved global list into helper file

ZephyrTFA committed:
remove 'intercept' replace with syndicate
because how would anyone know what intercept meant

ZephyrTFA committed:
add syndicate mainframe to cc to ensure syndicate comms always work

ZephyrTFA commented:
added syndicate mainframe to CC, also makes it to where people cant gamer WireCarp to see if there are nukies.
Also, someone should probably make wirecarp respect off network devices

ZephyrTFA committed:
forgot advanced infil

Mothblocks commented on code/controllers/subsystem/atoms.dm:
```diff
@@ -167,9 +167,15 @@ SUBSYSTEM_DEF(atoms)
 	set_tracked_initalized(INITIALIZATION_INSSATOMS)
+	INIT_COST(GLOB.post_init_cost, GLOB.post_init_count)
+	SET_COST("[time_stamp()]: SET another tracked initailized")
+	EXPORT_STATS_TO_FILE_LATER("lemon.log", GLOB.post_init_cost, GLOB.post_init_count)
```
gut these before test merge

Mothblocks replied:
oh i think you did

san7890 commented:
![image](https://user-images.githubusercontent.com/34697715/208793738-4964bdcb-ecb9-4943-b576-a28b6ad5f641.png)

Not a fan of the verb placement, feels more for like the debug tab? Unsure.

LemonInTheDark commented:
Fix is merged into this pr

ZephyrTFA commented:
main issue work putting it in mapping verbs proper is that it will require you to have R_DEBUG, and i only care about R_ADMIN

LemonInTheDark commented:
Honestly it's hard to event without vv, and so debug.
If people are so bad at hiring admins that they don't give them debug then I guess it sucks to be them

Mothblocks commented:
@ZephyrTFA Can you move it to events or debug and then I can merge this

ZephyrTFA committed:
Update admin_verbs.dm

ZephyrTFA commented:
Instructions unclear, new tab created explicitly for verb
